File: agrotech-back/src/main/java/br/com/agrotech/agrotech_api/service/DispositivoService.java
package br.com.agrotech.agrotech_api.service;

import br.com.agrotech.agrotech_api.model.Dispositivo;
import br.com.agrotech.agrotech_api.repository.DispositivoRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

/**
 * Classe de serviço para gerenciamento das regras de negócio dos dispositivos IoT.
 */
@Service
public class DispositivoService {

    @Autowired
    private DispositivoRepository dispositivoRepository;

    /**
     * Lista todos os dispositivos cadastrados no ecossistema da aplicação.
     *
     * @return Uma lista {@link List} contendo todos os objetos {@link Dispositivo}.
     */
    public List<Dispositivo> buscarTodos() {
        return dispositivoRepository.findAll();
    }

    /**
     * Localiza um equipamento IoT cadastrado a partir de seu identificador principal.
     *
     * @param id Identificador único do dispositivo.
     * @return Um {@link Optional} contendo o {@link Dispositivo} encontrado.
     */
    public Optional<Dispositivo> buscarPorId(Long id) {
        return dispositivoRepository.findById(id);
    }

    /**
     * Salva um novo dispositivo no banco de dados.
     *
     * @param dispositivo A instância do dispositivo contendo os dados de instalação e status.
     * @return O {@link Dispositivo} persistido com seu ID preenchido.
     */
    public Dispositivo salvar(Dispositivo dispositivo) {
        return dispositivoRepository.save(dispositivo);
    }

    /**
     * Atualiza as informações de registro, status ou associação de fazenda de um dispositivo IoT.
     *
     * @param id Identificador do dispositivo a ser modificado.
     * @param dispositivoAtualizado Objeto com os dados atualizados do equipamento.
     * @return O {@link Dispositivo} atualizado após o salvamento no banco.
     * @throws RuntimeException Se o equipamento informado não existir no banco.
     */
    public Dispositivo atualizar(Long id, Dispositivo dispositivoAtualizado) {
        return dispositivoRepository.findById(id)
                .map(dispositivo -> {
                    dispositivo.setNomeModelo(dispositivoAtualizado.getNomeModelo());
                    dispositivo.setDataInstalacao(dispositivoAtualizado.getDataInstalacao());
                    dispositivo.setStatus(dispositivoAtualizado.getStatus());
                    if (dispositivoAtualizado.getFazenda() != null) {
                        dispositivo.setFazenda(dispositivoAtualizado.getFazenda());
                    }
                    return dispositivoRepository.save(dispositivo);
                })
                .orElseThrow(() -> new RuntimeException("Dispositivo não encontrado com o ID: " + id));
    }

    /**
     * Remove um equipamento IoT do cadastro ativo da aplicação.
     *
     * @param id Identificador do dispositivo a ser deletado.
     * @throws RuntimeException Se o equipamento informado não for localizado.
     */
    public void deletar(Long id) {
        if (!dispositivoRepository.existsById(id)) {
            throw new RuntimeException("Dispositivo não encontrado com o ID: " + id);
        }
        dispositivoRepository.deleteById(id);
    }
}

File: agrotech-front/src/services/dispositivoService.ts
import { api } from './api';
import { Dispositivo } from '../types/Dispositivo';

/**
 * Serviço responsável por realizar as chamadas HTTP relacionadas
 * ao gerenciamento e controle dos dispositivos IoT.
 */
export const dispositivoService = {
    /**
     * Consulta a lista com todos os dispositivos IoT registrados.
     *
     * @returns {Promise<Dispositivo[]>} Uma promessa contendo a lista de equipamentos.
     */
    listarTodos: async (): Promise<Dispositivo[]> => {
        const response = await api.get<Dispositivo[]>('/dispositivos');
        return response.data;
    },

    /**
     * Localiza um dispositivo IoT específico pelo seu ID.
     *
     * @param {number} id - Identificador único do dispositivo.
     * @returns {Promise<Dispositivo>} Uma promessa contendo os dados do dispositivo.
     */
    buscarPorId: async (id: number): Promise<Dispositivo> => {
        const response = await api.get<Dispositivo>(`/dispositivos/${id}`);
        return response.data;
    },

    /**
     * Registra um novo equipamento IoT vinculando-o a uma fazenda.
     *
     * @param {Dispositivo} dispositivo - Objeto contendo dados de modelo, status e vínculo.
     * @returns {Promise<Dispositivo>} Uma promessa com o dispositivo salvo.
     */
    cadastrar: async (dispositivo: Dispositivo): Promise<Dispositivo> => {
        const response = await api.post<Dispositivo>('/dispositivos', dispositivo);
        return response.data;
    },

    /**
     * Altera informações ou o status operacional de um equipamento cadastrado.
     *
     * @param {number} id - Identificador do dispositivo.
     * @param {Dispositivo} dispositivo - Novos dados para atualização.
     * @returns {Promise<Dispositivo>} Uma promessa contendo o dispositivo atualizado.
     */
    atualizar: async (id: number, dispositivo: Dispositivo): Promise<Dispositivo> => {
        const response = await api.put<Dispositivo>(`/dispositivos/${id}`, dispositivo);
        return response.data;
    },

    /**
     * Exclui o cadastro de um equipamento do sistema.
     *
     * @param {number} id - Identificador único do dispositivo a ser deletado.
     * @returns {Promise<void>} Promessa resolvida após a exclusão.
     */
    deletar: async (id: number): Promise<void> => {
        await api.delete(`/dispositivos/${id}`);
    },
};